rust/arrow/src/array/builder.rs

```diff
@@ -973,6 +995,105 @@ impl Drop for StructBuilder {
     }
 }
 
+/// Array builder for `DictionaryArray`. For example to map a set of byte indices
+/// to f32 values. Note that the use of a `HashMap` here will not scale to very large
+/// arrays or result in an ordered dictionary.
+pub struct PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    keys_builder: PrimitiveBuilder<K>,
+    values_builder: PrimitiveBuilder<V>,
+    map: HashMap<&'static [u8], K::Native>,
+}
+
+impl<K, V> PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Creates a new `PrimitiveDictionaryBuilder` from a keys builder and a value builder.
+    pub fn new(
+        keys_builder: PrimitiveBuilder<K>,
+        values_builder: PrimitiveBuilder<V>,
+    ) -> Self {
+        Self {
+            keys_builder: keys_builder,
+            values_builder: values_builder,
+            map: HashMap::new(),
+        }
+    }
+}
+
+impl<K, V> ArrayBuilder for PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Returns the builder as an non-mutable `Any` reference.
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    /// Returns the builder as an mutable `Any` reference.
+    fn as_any_mut(&mut self) -> &mut Any {
+        self
+    }
+
+    /// Returns the boxed builder as a box of `Any`.
+    fn into_box_any(self: Box<Self>) -> Box<Any> {
+        self
+    }
+
+    /// Returns the number of array slots in the builder
+    fn len(&self) -> usize {
+        self.keys_builder.len()
+    }
+
+    /// Builds the array and reset this builder.
+    fn finish(&mut self) -> ArrayRef {
+        Arc::new(self.finish())
+    }
+}
+
+impl<K, V> PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Append a primive value to the array. Return an existing index
+    /// if already present in the values array or a new index if the
+    /// value is appended to the values array.
+    /// ```
+    /// ```
+    pub fn append(&mut self, value: V::Native) -> Result<K::Native> {
+        if let Some(&key) = self.map.get(value.to_byte_slice()) {
+            // Append existing value.
+            self.keys_builder.append_value(key)?;
+            Ok(key)
+        } else {
+            // Append new value.
+            let key = K::Native::from_usize(self.values_builder.len())
+                .ok_or(ArrowError::DictionaryKeyOverflowError)?;
+            self.values_builder.append_value(value)?;
+            self.keys_builder.append_value(key as K::Native)?;
+            Ok(key)
+        }
+    }
+
+    pub fn append_null(&mut self) -> Result<()> {
```

Comment:
There are two ways to have nulls in the dictionaries.
1) Have a null in the values (say at index 0) and use a special index.
    Effective if the length of the values array is not already 256 or 65536 so adding an extra
    value will use many more bytes.
2) Have a null vector in the keys array.
    Effective if the length of the values array is exactly 256 or 65536.

I was proposing having an "add_special_null" function to designate a null element
in the values array or a boolean to describe the 0 index null scheme. For SIMD
code, having 0 as the null fits well with the AVX512 conditional execution, for example.

To be honest, if I were the user, I would much prefer to have full control over the key
and value arrays for performance sake, ie. build the values first myself and construct
the dictionary from the two arrays.

Reply:
I don't think we need a special index because we have the null buffer for that. I'll try to create some examples on a fork of your branch.

> To be honest, if I were the user, I would much prefer to have full control over the key
and value arrays for performance sake, ie. build the values first myself and construct
the dictionary from the two arrays.

We ca still make this possible in follow-up PRs by creating constructors that take in 2 arrays (keys and values). The `DictionaryArray::from()` also covers this use-case, in fact that's what I'd expect users would naturally use to construct dictionary arrays